micrometer-core/src/test/java/io/micrometer/core/instrument/binder/okhttp3/OkHttpMetricsEventListenerTest.java

```diff
@@ -141,4 +145,25 @@ void uriTagWorksWithUriMapper(@WiremockResolver.Wiremock WireMockServer server)
                 .tags("foo", "bar", "uri", "/helloworld.txt", "status", "200")
                 .timer().count()).isEqualTo(1L);
     }
+
+    @Test
+    void cachedRequestsDoNotLeakMemory(@WiremockResolver.Wiremock WireMockServer server, @TempDir Path tempDir) throws IOException {
+        OkHttpMetricsEventListener okHttpMetricsEventListener = OkHttpMetricsEventListener.builder(registry, "okhttp.requests").build();
+        OkHttpClient clientWithCache = new OkHttpClient.Builder()
+                .eventListener(okHttpMetricsEventListener)
+                .cache(new Cache(tempDir.toFile(), 55555))
+                .build();
+        server.stubFor(any(anyUrl()).willReturn(aResponse().withHeader("Cache-Control", "max-age=9600")));
+        Request request = new Request.Builder()
+                .url(server.baseUrl())
+                .build();
+
+        clientWithCache.newCall(request).execute().close();
+        assertThat(okHttpMetricsEventListener.callState).isEmpty();
+        try (Response response = clientWithCache.newCall(request).execute()) {
+            assertThat(response.cacheResponse()).isNotNull();
+        }
+
+        assertThat(okHttpMetricsEventListener.callState).isEmpty();
```

Comment:
I added this test which I think captures the scenario described. Unfortunately, this assertion is failing for me. Any ideas why @ahatzz11? I'm not seeing the `callEnd` callback called when closing the cached response.

Reply:
@shakuzen Looking at the test my first guess was that the second request that is made is saved to `Response response` on line 163, but that response never has a `response.close()`. 

I pulled your test down and it's passing for me locally without any changes, not sure why that is yet. 

I'm curious - why is there a `try` block here?

Reply:
The try block is what closes the Response when the block closes.

Reply:
I've done some debugging locally on our project and all of our cached calls end up hitting the `callEnd()` method and the `callState` becomes empty. 

To me this test is accurately testing this scenario. It seems like there is a call that is still in the `callState` despite both calls being closed out which would is partly what this PR is trying to resolve. 

I'll continue digging more tomorrow.

Reply:
The test is failing on the 1.1.x branch because it's using OkHttp 3.11 which has a bug in which `callEnd` was not executed. This was fixed around the `3.14.2` version of OkHttp:
https://github.com/square/okhttp/commit/59e68cd2bcba789e7faca70f8910afd2edeed2be
https://github.com/square/okhttp/issues/5143

@shakuzen how do you want to address this? Is upgrading okhttp in a maintenance branch acceptable? Alternatively, we could use the `WeakHashMap` change I described above and then include a release note that says when using the metrics integration, to be sure to upgrade to okhttp 3.14.2 or newer.

Reply:
@johnrengelman thank you for tracking this down. Since we only have an optional dependency on OkHttp, the version is for checking compatibility and that the integration works as expected back to that version. We typically don't upgrade minor versions of dependencies in maintenance branches, so we can ensure the code works with versions of the library from that generation. That said, we could make an exception and upgrade given a fix isn't available in `3.11.x`.

I wanted to consider the change to a `WeakHashMap` separate from this fix, and probably only in `master`. I worry a bit that it would mask this kind problem where we should be explicitly cleaning up something but are not.

Reply:
@shakuzen dilemma understood! Do you want us to leave this in your hands to figure out or is there something we can do to help move this forward?

Reply:
I also verified that no other methods on `EventListener` are invoked for a cached response in v3.11 of OkHttp.

Reply:
I think I found a solution, but I think it still runs into a compatibility issue. We could change `OkHttpMetricsEventListener` to extend `EventListener.Factory` instead of `EventListener`. Then OkHttp would create a new listener per `Call` instead of a single global  listener which would get GC'd with each Call.

Reply:
I've gone ahead and upgraded the okhttp dependency to 3.14.7 and the test is passing now. We can include a note in the release note about this. From the information shared here, it's only critical (though obviously recommended for everyone) for users using a `Cache` with okhttp to upgrade for this fix to work, correct?

Reply:
That seems to be the case, yes.

Reply:
Since `callEnd()` wasn't implemented before wouldn't it also be possible to have a call get stuck if anything other than an `IOException` was thrown?